database/database.go

```diff
@@ -15,13 +20,79 @@ const (
 )
 
 var (
-	ErrKeyNotFound              = errors.New("key not found")
-	ErrChainNotFound            = errors.New("no database for chain")
+	ErrDataKeyNotFound          = errors.New("data key not found")
+	ErrRelayerKeyNotFound       = errors.New("no database for relayer key")
 	ErrDatabaseMisconfiguration = errors.New("database misconfiguration")
 )
 
 // RelayerDatabase is a key-value store for relayer state, with each blockchainID maintaining its own state
 type RelayerDatabase interface {
-	Get(blockchainID ids.ID, key []byte) ([]byte, error)
-	Put(blockchainID ids.ID, key []byte, value []byte) error
+	Get(relayerKey common.Hash, dataKey []byte) ([]byte, error)
+	Put(relayerKey common.Hash, dataKey []byte, value []byte) error
+}
+
+// Returns true if an error returned by a RelayerDatabase indicates the requested key was not found
+func IsKeyNotFoundError(err error) bool {
+	return errors.Is(err, ErrRelayerKeyNotFound) || errors.Is(err, ErrDataKeyNotFound)
+}
+
+// RelayerKey is a unique identifier for an application relayer
+type RelayerKey struct {
+	SourceBlockchainID      ids.ID
+	DestinationBlockchainID ids.ID
+	OriginSenderAddress     common.Address
+	DestinationAddress      common.Address
+}
+
+// CalculateRelayerKey calculates the unique identifier for an application relayer
+func (k RelayerKey) CalculateRelayerKey() common.Hash {
+	return CalculateRelayerKey(
+		k.SourceBlockchainID,
+		k.DestinationBlockchainID,
+		k.OriginSenderAddress,
+		k.DestinationAddress,
+	)
+}
+
+// Standalone utility to calculate a relayer key
+func CalculateRelayerKey(
+	sourceBlockchainID ids.ID,
+	destinationBlockchainID ids.ID,
+	originSenderAddress common.Address,
+	desinationAddress common.Address,
+) common.Hash {
+	return crypto.Keccak256Hash(
+		[]byte(strings.Join(
+			[]string{
+				sourceBlockchainID.String(),
+				destinationBlockchainID.String(),
+				originSenderAddress.String(),
+				desinationAddress.String(),
+			},
+			"-",
+		)),
+	)
+}
+
+// Get all of the possible relayer keys for a given configuration
+func GetConfigRelayerKeys(cfg *config.Config) []RelayerKey {
+	var keys []RelayerKey
+	for _, s := range cfg.SourceBlockchains {
+		keys = append(keys, GetSourceConfigRelayerKeys(s)...)
+	}
+	return keys
+}
+
+// Calculate all of the possible relayer keys for a given source blockchain
+func GetSourceConfigRelayerKeys(cfg *config.SourceBlockchain) []RelayerKey {
```

Comment:
nit: can this be renamed to `GetSourceBlockchainRelayerKeys`? and possibly renaming argument parameter too

Reply:
does this function and the underlying functions it calls need to take in a pointer?

Reply:
> does this function and the underlying functions it calls need to take in a pointer?

`Config.GetSupportedDestinations` takes a pointer receiver, so the pointer arguments just prevent us from having to make that conversion in this call.

I'm not sure if this is what you were thinking about, but I'd definitely like to take steps to increase the immutability of the `Config` singleton. In the absence of const pointers, we'd have to refactor the `Config` type itself a bit to provide tighter access control.